src/complexity_measures/sample_entropy.jl

```diff
@@ -0,0 +1,144 @@
+using DelayEmbeddings
+using Statistics
+using Neighborhood.NearestNeighbors: inrangecount, Chebyshev
+
+export SampleEntropy
+
+"""
+    SampleEntropy(; r::Real = 0.2 * Statistics.std(x), m::Int = 2, τ::Int = 1,
+        metric = Chebyshev()
+    SampleEntropy(x::AbstractVector; m = 2, τ = 1, metric = Chebyshev())
+
+An estimator for the sample entropy complexity measure (Richman & Moorman,
+2000)[^Richman2000], used with [`complexity`](@ref) and [`complexity_normalized`](@ref).
+
+Providing a univariate timeseries `x` to the constructors automatically determines the
+radius `r` as `0.2 * Statistics.std(x)`.
+
+## Description
+
+Sample entropy is defined as
+
+```math
+SampEn(m, r) = \\lim_{N \\to \\infty} \\left[ -\\ln \\dfrac{A^{m+1}(r)}{B^m(r)} \\right],
+```
+
+An *estimator* for sample entropy using radius `r`, embedding dimension `m`,
+embedding lag `τ` is
+
+```math
+SampEn(m,r, N) = -\\ln{\\dfrac{A(r, N)}{B(r, N)}}.
+```
+
+Here,
+
+```math
+\\begin{aligned}
+B(r, m, N) = \\sum_{i = 1}^{N-m\\tau} \\sum_{j = 1, j \\neq i}^{N-m\\tau} \\theta(d({\\bf x}_i^m, {\\bf x}_j^m) \\leq r) \\\\
+A(r, m, N) = \\sum_{i = 1}^{N-m\\tau} \\sum_{j = 1, j \\neq i}^{N-m\\tau} \\theta(d({\\bf x}_i^{m+1}, {\\bf x}_j^{m+1}) \\leq r) \\\\
+\\end{aligned},
+```
```

Comment:
Hmmm this is supsiciously similar to Cao's method for deducing an optimal embedding... Right? https://juliadynamics.github.io/DynamicalSystems.jl/dev/embedding/traditional/#DelayEmbeddings.delay_afnn you don't compute the average distance, but you compute how many are within distance `r`, and how this changes from embedding `m` to embedding `m+1`.

Reply:
I'm not intimately familiar with Cao's method, but just had a quick glance at the source code. It seems there is some overlap. I see that you use `bulkisearch` in the `_average_a` function. I saw at least a 2x speed up and 2+ time less allocations here by using `inrangecount`, so using it there too could probably improve performance.